contracts/zero-ex/contracts/src/features/LimitOrdersFeature.sol

```diff
@@ -0,0 +1,875 @@
+                msg.sender
+            );
+        _refundProtocolFeeToSender(ethProtocolFeePaid);
+    }
+
+    /// @dev Fill an RFQ order for exactly `takerTokenFillAmount` taker tokens.
+    ///      The taker will be the caller. ETH protocol fees can be
+    ///      attached to this call. Any unspent ETH will be refunded to
+    ///      the caller.
+    /// @param order The limit order.
+    /// @param signature The order signature.
+    /// @param takerTokenFillAmount How much taker token to fill this order with.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function fillOrKillLimitOrder(
+        LibLimitOrder.LimitOrder memory order,
+        LibSignature.Signature memory signature,
+        uint128 takerTokenFillAmount
+    )
+        public
+        override
+        payable
+        returns (uint256 makerTokenFilledAmount)
+    {
+        uint256 ethProtocolFeePaid;
+        uint256 takerTokenFilledAmount;
+        (ethProtocolFeePaid, takerTokenFilledAmount, makerTokenFilledAmount) =
+            _fillLimitOrderPrivate(FillLimitOrderPrivateParams({
+                order: order,
+                signature: signature,
+                takerTokenFillAmount: takerTokenFillAmount,
+                taker: msg.sender,
+                sender: msg.sender
+            }));
+        // Must have filled exactly the amount requested.
+        if (takerTokenFilledAmount != takerTokenFillAmount) {
+            LibLimitOrdersRichErrors.FillOrKillFailedError(
+                getLimitOrderHash(order),
+                takerTokenFilledAmount,
+                takerTokenFillAmount
+            ).rrevert();
+        }
+        _refundProtocolFeeToSender(ethProtocolFeePaid);
+    }
+
+    /// @dev Fill an RFQ order for exactly `takerTokenFillAmount` taker tokens.
+    ///      The taker will be the caller. ETH protocol fees can be
+    ///      attached to this call. Any unspent ETH will be refunded to
+    ///      the caller.
+    /// @param order The RFQ order.
+    /// @param signature The order signature.
+    /// @param takerTokenFillAmount How much taker token to fill this order with.
+    /// @return makerTokenFilledAmount How much maker token was filled.
+    function fillOrKillRfqOrder(
+        LibLimitOrder.RfqOrder memory order,
+        LibSignature.Signature memory signature,
+        uint128 takerTokenFillAmount
+    )
+        public
+        override
+        payable
+        returns (uint256 makerTokenFilledAmount)
+    {
+        uint256 ethProtocolFeePaid;
+        uint256 takerTokenFilledAmount;
+        (ethProtocolFeePaid, takerTokenFilledAmount, makerTokenFilledAmount) =
+            _fillRfqOrderPrivate(
+                order,
+                signature,
+                takerTokenFillAmount,
+                msg.sender
+            );
+        // Must have filled exactly the amount requested.
+        if (takerTokenFilledAmount != takerTokenFillAmount) {
```

Comment:
Thinking about rounding errors, it could be tricky to get exactly `takerTokenFilledAmount`. Perhaps we could be a bit more lenient and do `if (takerTokenFilledAmount < takerTokenFillAmount)`? (And above)